metrics/metrics.go

```diff
@@ -101,26 +208,105 @@ var _ CSIMetricsManager = &csiMetricsManager{}
 
 type csiMetricsManager struct {
 	registry                   metrics.KubeRegistry
+	subsystem                  string
+	stabilityLevel             metrics.StabilityLevel
 	driverName                 string
-	csiOperationsMetric        *metrics.CounterVec
+	additionalLabelNames       []string
+	additionalLabels           []label
 	csiOperationsLatencyMetric *metrics.HistogramVec
 }
 
+type label struct {
+	name, value string
+}
+
 func (cmm *csiMetricsManager) GetRegistry() metrics.KubeRegistry {
 	return cmm.registry
 }
 
-// RecordMetrics must be called upon CSI Operation completion to record
-// the operation's metric.
-// operationName - Name of the CSI operation.
-// operationErr - Error, if any, that resulted from execution of operation.
-// operationDuration - time it took for the operation to complete
+// RecordMetrics implements CSIMetricsManager.RecordMetrics.
 func (cmm *csiMetricsManager) RecordMetrics(
 	operationName string,
 	operationErr error,
 	operationDuration time.Duration) {
-	cmm.csiOperationsLatencyMetric.WithLabelValues(
-		cmm.driverName, operationName, getErrorCode(operationErr)).Observe(operationDuration.Seconds())
+	cmm.recordMetricsWithLabels(operationName, operationErr, operationDuration)
+}
+
+// recordMetricsWithLabels is the internal implementation of RecordMetrics.
+func (cmm *csiMetricsManager) recordMetricsWithLabels(
+	operationName string,
+	operationErr error,
+	operationDuration time.Duration,
+	labelValues ...string) {
+	values := []string{cmm.driverName, operationName, getErrorCode(operationErr)}
+	toAdd := len(labelValues)
+	if toAdd > len(cmm.additionalLabelNames) {
```

Comment:
I think we can avoid all of this logic if we just pass in the labels as a map generated in WithLabelValues()

Reply:
You are right, that makes the code simpler. Change pushed.